File: native-views/react-native-native-list/ios/NativeListActivityDetailsView.swift
import UIKit

/// The activity template's bounded rich presentation; all values are caller-formatted.
final class NativeListActivityDetailsView: UIStackView {
  var onAction: ((String, UIView, String, Int) -> Void)?
  private let main = UIStackView()
  private let visual = NativeListLeadingVisual()
  private let identity = UIStackView()
  private let titleLine = UIStackView()
  private let title = NativeListTextLabel()
  private let descriptionLabel = NativeListTextLabel()
  private let status = NativeListTextLabel()
  private let amounts = UIStackView()
  private let fee = UIStackView()
  private let feeLabel = NativeListTextLabel()
  private let feeLine = UIStackView()
  private let feePrimary = NativeListTextLabel()
  private let feeSecondary = NativeListTextLabel()
  private let actions = UIStackView()
  private var amountViews: [NativeListActivityAmountView] = []
  private var badgeViews: [NativeListTextLabel] = []
  private var buttons: [UIButton] = []
  private var actionKeys: [String] = []
  private var descriptionKey = ""
  private var dimensions: [NSLayoutConstraint] = []

  override init(frame: CGRect) {
    super.init(frame: frame)
    axis = .vertical; alignment = .fill; spacing = 8
    main.axis = .horizontal; main.alignment = .center; main.spacing = 12
    identity.axis = .vertical; identity.alignment = .fill
    titleLine.axis = .horizontal; titleLine.alignment = .center; titleLine.spacing = 4
    titleLine.addArrangedSubview(title)
    [titleLine, descriptionLabel, status].forEach(identity.addArrangedSubview)
    amounts.axis = .vertical; amounts.alignment = .fill; amounts.spacing = 2
    fee.axis = .vertical; fee.alignment = .trailing
    feeLine.axis = .horizontal; feeLine.spacing = 4
    [feePrimary, feeSecondary].forEach(feeLine.addArrangedSubview)
    [feeLabel, feeLine].forEach(fee.addArrangedSubview)
    [visual, identity, amounts, fee].forEach(main.addArrangedSubview)
    [main, actions].forEach(addArrangedSubview)
    actions.axis = .horizontal; actions.alignment = .leading; actions.spacing = 8
    actions.isLayoutMarginsRelativeArrangement = true
    descriptionLabel.isUserInteractionEnabled = true
    descriptionLabel.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(descriptionPressed)))
    title.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
    identity.setContentHuggingPriority(.defaultLow, for: .horizontal)
    amounts.setContentHuggingPriority(.defaultHigh, for: .horizontal)
  }
  required init(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

  static func measure(_ data: [String: Any], width: CGFloat) -> CGFloat {
    let style = data.dictionary("style") ?? [:]
    let primary = style.dictionary("primaryAmount") ?? [:]
    let secondary = style.dictionary("secondaryAmount") ?? [:]
    let rows = data.dictionaries("amounts")
    let gap = CGFloat(style.double("trailingGap", default: 2))
    let amountsHeight = rows.reduce(CGFloat(0)) { height, row in
      height + CGFloat(primary.double("lineHeight", default: 24)) +
        (row.string("secondaryText").isEmpty ? 0 : CGFloat(secondary.double("lineHeight", default: 20)))
    } + CGFloat(max(0, rows.count - 1)) * gap
    let titleHeight = CGFloat(style.dictionary("title")?.double("lineHeight", default: 24) ?? 24)
    let description = NativeListResolvedText(data.string("description"), style: style.dictionary("description"),
      size: 14, color: .label, lineHeight: 20, lines: 2)
    let identityHeight = titleHeight + description.measure(width: max(1, width * 0.4 - 52)) +
      (data.string("status").isEmpty ? 0 : 16)
    let imageHeight = CGFloat(style.dictionary("image")?.double("height", default: 40) ?? 40)
    return max(imageHeight, identityHeight, amountsHeight) +
      CGFloat(style.double("verticalPadding", default: 8)) * 2 +
      (data.dictionaries("footerActions").isEmpty ? 0 : 40)
  }

  func bind(_ item: NativeListItem, theme: [String: Any]?) {
    let data = item.data
    let style = data.dictionary("style") ?? [:]
    let image = style.dictionary("image") ?? [:]
    let table = data.string("presentation") == "table"
    let gap = CGFloat(style.double("leadingGap", default: 12))
    main.spacing = gap
    identity.spacing = CGFloat(style.double("lineGap"))
    amounts.spacing = CGFloat(style.double("trailingGap", default: 2))
    titleLine.spacing = CGFloat(style.double("titleBadgeGap", default: 4))
    NSLayoutConstraint.deactivate(dimensions)
    dimensions = [
      visual.widthAnchor.constraint(equalToConstant: CGFloat(image.double("width", default: 40))),
      visual.heightAnchor.constraint(equalToConstant: CGFloat(image.double("height", default: 40))),
      amounts.widthAnchor.constraint(lessThanOrEqualTo: main.widthAnchor, multiplier: table ? 0.42 : 0.5),
    ]
    if table { dimensions.append(amounts.widthAnchor.constraint(equalTo: identity.widthAnchor)) }
    NSLayoutConstraint.activate(dimensions)
    visual.bind(data.dictionary("leading") ?? [:], style: image, key: item.key, theme: theme,
      isUnread: false, secondaryVisual: data.dictionary("secondaryLeading"))
    NativeListResolvedText(data.string("title"), style: style.dictionary("title"), size: 16,
      weight: .medium, color: nativeListColor(theme, "primaryText", "#202020"), lineHeight: 24, lines: 1).bind(title)
    NativeListResolvedText(data.string("description"), style: style.dictionary("description"), size: 14,
      color: nativeListColor(theme, "secondaryText", "#646464"), lineHeight: 20, lines: 2).bind(descriptionLabel)
    NativeListResolvedText(data.string("status"), style: style.dictionary("status"), size: 12,
      color: nativeListColor(theme, "secondaryText", "#646464"), lineHeight: 16, lines: 1).bind(status)
    descriptionKey = data.string("descriptionActionKey")
    descriptionLabel.isUserInteractionEnabled = !descriptionKey.isEmpty
    descriptionLabel.accessibilityTraits = descriptionKey.isEmpty ? .staticText : .button
    for badge in badgeViews { titleLine.removeArrangedSubview(badge); badge.removeFromSuperview() }
    badgeViews = data.dictionaries("badges").map { descriptor in
      let label = NativeListTextLabel()
      let tone = descriptor.string("tone")
      let tint = nativeListActivityTone(tone, theme: theme)
      NativeListResolvedText("  \(descriptor.string("text"))  ", style: nil, size: 12,
        weight: .medium, color: tint, lineHeight: 16, lines: 1).bind(label)
      label.backgroundColor = tint.withAlphaComponent(0.1)
      label.layer.cornerRadius = 4; label.clipsToBounds = true
      label.setContentCompressionResistancePriority(.required, for: .horizontal)
      titleLine.addArrangedSubview(label)
      return label
    }
    let descriptors = data.dictionaries("amounts")
    while amountViews.count > descriptors.count {
      let view = amountViews.removeLast(); view.recycle(); amounts.removeArrangedSubview(view); view.removeFromSuperview()
    }
    while amountViews.count < descriptors.count {
      let view = NativeListActivityAmountView(); amountViews.append(view); amounts.addArrangedSubview(view)
    }
    for (index, descriptor) in descriptors.enumerated() {
      amountViews[index].bind(descriptor, key: "\(item.key):\(descriptor.string("key"))", style: style, theme: theme, table: table)
    }
    let feeData = data.dictionary("fee")
    fee.isHidden = feeData == nil
    fee.alpha = feeData?.bool("hidden") == true ? 0 : 1
    NativeListResolvedText(feeData?.string("label") ?? "", style: style.dictionary("secondaryAmount"),
      size: 12, color: nativeListColor(theme, "secondaryText", "#646464"), lineHeight: 16, lines: 1).bind(feeLabel)
    NativeListResolvedText(feeData?.string("primary") ?? "", style: style.dictionary("secondaryAmount"),
      size: 14, color: nativeListColor(theme, "primaryText", "#202020"), lineHeight: 20, lines: 1).bind(feePrimary)
    NativeListResolvedText(feeData?.string("secondary") ?? "", style: style.dictionary("secondaryAmount"),
      size: 12, color: nativeListColor(theme, "secondaryText", "#646464"), lineHeight: 16, lines: 1).bind(feeSecondary)
    for button in buttons { actions.removeArrangedSubview(button); button.removeFromSuperview() }
    buttons = []
    nativeListActivitySegments(feePrimary, feeData?.dictionaries("primaryTextSegments") ?? [])
    nativeListActivitySegments(feeSecondary, feeData?.dictionaries("secondaryTextSegments") ?? [])
    let actionData = data.dictionaries("footerActions")
    actionKeys = actionData.map { $0.string("key") }
    actions.isHidden = actionData.isEmpty
    actions.directionalLayoutMargins = NSDirectionalEdgeInsets(top: 0, leading: CGFloat(image.double("width", default: 40)) + gap, bottom: 0, trailing: 0)
    for (index, descriptor) in actionData.enumerated() {
      let button = UIButton(type: .system); button.tag = index
      button.setTitle(descriptor.string("label"), for: .normal)
      button.titleLabel?.font = nativeListFont(ofSize: 14, weight: .medium)
      button.setTitleColor(nativeListActivityTone(descriptor.string("tone"), theme: theme), for: .normal)
      button.backgroundColor = nativeListColor(theme, "strongBackground", "#F0F0F0")
      button.layer.cornerRadius = 8
      button.contentEdgeInsets = UIEdgeInsets(top: 4, left: 8, bottom: 4, right: 8)
      button.isEnabled = !data.bool("disabled") && !descriptor.bool("disabled")
      button.addTarget(self, action: #selector(actionPressed(_:)), for: .touchUpInside)
      buttons.append(button); actions.addArrangedSubview(button)
    }
    let spacer = UIView(); spacer.setContentHuggingPriority(.defaultLow, for: .horizontal)
    // The last flexible spacer is separate from the fixed action buttons.
    actions.arrangedSubviews.filter { !($0 is UIButton) }.forEach { actions.removeArrangedSubview($0); $0.removeFromSuperview() }
    actions.addArrangedSubview(spacer)
  }
  @objc private func descriptionPressed() { onAction?(descriptionKey, descriptionLabel, "description", 0) }
  @objc private func actionPressed(_ button: UIButton) {
    guard button.tag < actionKeys.count else { return }
    onAction?(actionKeys[button.tag], button, "footerAction", button.tag)
  }
  func recycle() {
    visual.recycle(); amountViews.forEach { $0.recycle() }; actionKeys = []; descriptionKey = ""
  }
}

private func nativeListActivityTone(_ tone: String, theme: [String: Any]?) -> UIColor {
  switch tone {
  case "positive", "success": return nativeListColor(theme, "positive", "#218358")
  case "negative", "danger": return nativeListColor(theme, "negative", "#CE2C31")
  case "secondary": return nativeListColor(theme, "secondaryText", "#646464")
  case "warning": return nativeListColor(theme, "warning", "#AB6400")
  case "info": return nativeListColor(theme, "info", "#007BEF")
  default: return nativeListColor(theme, "primaryText", "#202020")
  }
}

private final class NativeListActivityAmountView: UIStackView {
  private let visual = NativeListLeadingVisual()
  private let labels = UIStackView()
  private let primary = NativeListTextLabel()
  private let secondary = NativeListTextLabel()
  override init(frame: CGRect) {
    super.init(frame: frame)
    axis = .horizontal; alignment = .center; spacing = 6
    labels.axis = .vertical; labels.alignment = .fill
    [primary, secondary].forEach(labels.addArrangedSubview)
    [visual, labels].forEach(addArrangedSubview)
    NSLayoutConstraint.activate([visual.widthAnchor.constraint(equalToConstant: 20), visual.heightAnchor.constraint(equalToConstant: 20)])
  }
  required init(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
  func bind(_ data: [String: Any], key: String, style: [String: Any], theme: [String: Any]?, table: Bool) {
    let leading = data.dictionary("leading")
    visual.isHidden = leading == nil
    if let leading { visual.bind(leading, style: ["width": 20, "height": 20], key: key, theme: theme, isUnread: false) } else { visual.recycle() }
    var primaryStyle = style.dictionary("primaryAmount") ?? [:]
    if primaryStyle["alignment"] == nil { primaryStyle["alignment"] = table ? "start" : "end" }
    let resolved = NativeListResolvedText(data.string("text"), style: primaryStyle, size: 16,
      weight: .medium, color: nativeListActivityTone(data.string("tone"), theme: theme), lineHeight: 24, lines: 1, tabular: true)
    resolved.bind(primary)
    nativeListActivitySegments(primary, data.dictionaries("textSegments"))
    var secondaryStyle = style.dictionary("secondaryAmount") ?? [:]
    if secondaryStyle["alignment"] == nil { secondaryStyle["alignment"] = table ? "start" : "end" }
    NativeListResolvedText(data.string("secondaryText"), style: secondaryStyle, size: 14,
      color: nativeListColor(theme, "secondaryText", "#646464"), lineHeight: 20, lines: 1).bind(secondary)
    nativeListActivitySegments(secondary, data.dictionaries("secondaryTextSegments"))
  }
  func recycle() { visual.recycle() }
}

private func nativeListActivitySegments(_ label: NativeListTextLabel, _ segments: [[String: Any]]) {
  guard !segments.isEmpty else { return }
  let text = segments.map { $0.string("text") }.joined()
  let attributes = (label.attributedText?.length ?? 0) > 0
    ? label.attributedText!.attributes(at: 0, effectiveRange: nil)
    : [.font: label.font as Any, .foregroundColor: label.textColor as Any]
  let value = NSMutableAttributedString(string: text, attributes: attributes)
  var offset = 0
  for segment in segments {
    let length = (segment.string("text") as NSString).length
    if segment.string("style") == "subscript" {
      value.addAttributes([.font: label.font.withSize(ceil(label.font.pointSize * 0.6))], range: NSRange(location: offset, length: length))
    }
    offset += length
  }
  label.attributedText = value
}

File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListActivityDetailsView.kt
package com.margelo.nitro.nativelist

import android.graphics.drawable.GradientDrawable
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.AbsoluteSizeSpan
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import com.facebook.react.uimanager.ThemedReactContext
import kotlin.math.roundToInt
import org.json.JSONObject

/** Bounded rich activity fields; no wallet or transaction semantics live here. */
internal class NativeListActivityDetailsView(private val reactContext: ThemedReactContext) : LinearLayout(reactContext) {
  var onAction: ((String, View, String, Int) -> Unit)? = null
  private var sourceScale = false
  private val main = LinearLayout(context).apply { orientation = HORIZONTAL; gravity = Gravity.CENTER_VERTICAL }
  private val visual = NativeListLeadingVisual(reactContext)
  private val identity = LinearLayout(context).apply { orientation = VERTICAL }
  private val titleLine = LinearLayout(context).apply { orientation = HORIZONTAL; gravity = Gravity.CENTER_VERTICAL }
  private val title = NativeListTextView(context)
  private val description = NativeListTextView(context)
  private val status = NativeListTextView(context)
  private val amounts = LinearLayout(context).apply { orientation = VERTICAL }
  private val fee = LinearLayout(context).apply { orientation = VERTICAL; gravity = Gravity.END }
  private val feeLabel = NativeListTextView(context)
  private val feeLine = LinearLayout(context).apply { orientation = HORIZONTAL; gravity = Gravity.END }
  private val feePrimary = NativeListTextView(context)
  private val feeSecondary = NativeListTextView(context)
  private val actions = LinearLayout(context).apply { orientation = HORIZONTAL }
  private val amountViews = mutableListOf<AmountView>()
  private val badges = mutableListOf<View>()

  init {
    orientation = VERTICAL
    titleLine.addView(title, LayoutParams(0, -2, 1f))
    identity.addView(titleLine)
    identity.addView(description)
    identity.addView(status)
    feeLine.addView(feePrimary); feeLine.addView(feeSecondary)
    fee.addView(feeLabel); fee.addView(feeLine)
    main.addView(visual)
    main.addView(identity, LayoutParams(0, -2, 1f))
    main.addView(amounts)
    main.addView(fee)
    addView(main, LayoutParams(-1, -2))
    addView(actions, LayoutParams(-1, -2))
  }
  private fun dp(value: Int) = if (sourceScale) (value * resources.displayMetrics.density).roundToInt() else NativeListScale.dp(resources, value)
  private fun px(value: Double) = (value * resources.displayMetrics.density).roundToInt()
  private fun color(tone: String, theme: JSONObject?): Int {
    val (key, fallback) = when (tone) {
      "positive", "success" -> "positive" to "#218358"
      "negative", "danger" -> "negative" to "#CE2C31"
      "secondary" -> "secondaryText" to "#646464"
      "warning" -> "warning" to "#AB6400"
      "info" -> "info" to "#007BEF"
      else -> "primaryText" to "#202020"
    }
    return parseNativeListColor(theme?.optString(key, fallback) ?: fallback)
  }
  private fun text(label: NativeListTextView, value: String, style: JSONObject?, size: Float,
                   ink: Int, line: Int, weight: String = "regular", lines: Int = 1) {
    NativeListResolvedText.resolve(context, value, style, size, weight, ink, line, lines, sourceScale).bind(label)
  }
  fun bind(item: NativeListItem, theme: JSONObject?, sourceScale: Boolean) {
    this.sourceScale = sourceScale
    val data = item.json
    val style = data.optJSONObject("style") ?: JSONObject()
    val image = style.optJSONObject("image") ?: JSONObject()
    val table = data.optString("presentation") == "table"
    val gap = if (style.has("leadingGap")) px(style.optDouble("leadingGap")) else dp(12)
    val iw = if (image.has("width")) px(image.optDouble("width")) else dp(40)
    visual.layoutParams = LayoutParams(iw, if (image.has("height")) px(image.optDouble("height")) else dp(40)).apply { marginEnd = gap }
    visual.bind(data.optJSONObject("leading") ?: JSONObject(), image, item.key, theme, false, sourceScale, data.optJSONObject("secondaryLeading"))
    amounts.layoutParams = if (table) LayoutParams(0, -2, 1f).apply { marginStart = gap } else LayoutParams(-2, -2).apply { marginStart = gap }
    text(title, data.optString("title"), style.optJSONObject("title"), 16f, color("primary", theme), 24, "medium")
    text(description, data.optString("description"), style.optJSONObject("description"), 14f, color("secondary", theme), 20, lines = 2)
    text(status, data.optString("status"), style.optJSONObject("status"), 12f, color("secondary", theme), 16)
    description.setOnClickListener(null)
    val descriptionKey = data.optString("descriptionActionKey")
    description.isClickable = descriptionKey.isNotEmpty()
    if (descriptionKey.isNotEmpty()) description.setOnClickListener { onAction?.invoke(descriptionKey, description, "description", 0) }
    for (view in listOf(description, status)) (view.layoutParams as LayoutParams).topMargin = if (style.has("lineGap")) px(style.optDouble("lineGap")) else 0
    badges.forEach(titleLine::removeView); badges.clear()
    val badgeData = data.optJSONArray("badges")
    for (index in 0 until (badgeData?.length() ?: 0)) {
      val descriptor = badgeData!!.getJSONObject(index)
      val label = NativeListTextView(context)
      val ink = color(descriptor.optString("tone"), theme)
      text(label, descriptor.optString("text"), null, 12f, ink, 16, "medium")
      label.setPadding(dp(4), 0, dp(4), 0)
      label.background = GradientDrawable().apply { setColor((ink and 0x00ffffff) or 0x19000000); cornerRadius = dp(4).toFloat() }
      titleLine.addView(label, LayoutParams(-2, -2).apply { marginStart = if (style.has("titleBadgeGap")) px(style.optDouble("titleBadgeGap")) else dp(4) })
      badges.add(label)
    }
    val amountData = data.optJSONArray("amounts")
    val count = amountData?.length() ?: 0
    while (amountViews.size > count) { val child = amountViews.removeAt(amountViews.lastIndex); child.recycle(); amounts.removeView(child) }
    while (amountViews.size < count) { val child = AmountView(); amountViews.add(child); amounts.addView(child) }
    for (index in 0 until count) {
      val descriptor = amountData!!.getJSONObject(index)
      amountViews[index].bind(descriptor, "${item.key}:${descriptor.optString("key")}", style, theme, table)
      amountViews[index].layoutParams = LayoutParams(-1, -2).apply { topMargin = if (index == 0) 0 else if (style.has("trailingGap")) px(style.optDouble("trailingGap")) else dp(2) }
    }
    val feeData = data.optJSONObject("fee")
    fee.visibility = if (feeData == null) GONE else VISIBLE
    fee.alpha = if (feeData?.optBoolean("hidden") == true) 0f else 1f
    fee.layoutParams = LayoutParams(-2, -2).apply { marginStart = gap }
    text(feeLabel, feeData?.optString("label") ?: "", style.optJSONObject("secondaryAmount"), 12f, color("secondary", theme), 16)
    text(feePrimary, feeData?.optString("primary") ?: "", style.optJSONObject("secondaryAmount"), 14f, color("primary", theme), 20)
    text(feeSecondary, feeData?.optString("secondary") ?: "", style.optJSONObject("secondaryAmount"), 12f, color("secondary", theme), 16)
    (feeSecondary.layoutParams as LayoutParams).marginStart = dp(4)
    applySegments(feePrimary, feeData?.optJSONArray("primaryTextSegments"), style.optJSONObject("secondaryAmount")?.optDouble("fontSize", 14.0) ?: 14.0)
    applySegments(feeSecondary, feeData?.optJSONArray("secondaryTextSegments"), style.optJSONObject("secondaryAmount")?.optDouble("fontSize", 12.0) ?: 12.0)
    actions.removeAllViews()
    val actionData = data.optJSONArray("footerActions")
    actions.visibility = if ((actionData?.length() ?: 0) == 0) GONE else VISIBLE
    actions.layoutParams = LayoutParams(-1, -2).apply { marginStart = iw + gap; topMargin = dp(8) }
    for (index in 0 until (actionData?.length() ?: 0)) {
      val descriptor = actionData!!.getJSONObject(index)
      val button = NativeListTextView(context)
      text(button, descriptor.optString("label"), null, 14f, color(descriptor.optString("tone"), theme), 20, "medium")
      button.setPadding(dp(8), dp(4), dp(8), dp(4))
      button.background = GradientDrawable().apply { setColor(parseNativeListColor(theme?.optString("strongBackground", "#F0F0F0") ?: "#F0F0F0")); cornerRadius = dp(8).toFloat() }
      button.isEnabled = !data.optBoolean("disabled") && !descriptor.optBoolean("disabled")
      button.setOnClickListener { onAction?.invoke(descriptor.optString("key"), button, "footerAction", index) }
      actions.addView(button, LayoutParams(-2, -2).apply { marginEnd = dp(8) })
    }
  }
  private fun applySegments(label: NativeListTextView, segments: org.json.JSONArray?, fontSize: Double) {
    if (segments == null || segments.length() == 0) return
    val fullText = (0 until segments.length()).joinToString("") { segments.getJSONObject(it).optString("text") }
    val styled = if (label.text.toString() == fullText) SpannableStringBuilder(label.text) else SpannableStringBuilder(fullText)
    var start = 0
    for (index in 0 until segments.length()) {
      val segment = segments.getJSONObject(index)
      val end = start + segment.optString("text").length
      if (segment.optString("style") == "subscript") {
        // Match NumberSizeableText: shrink the zero count without moving its baseline.
        val size = kotlin.math.ceil(fontSize * 0.6) * label.textSize / fontSize
        styled.setSpan(AbsoluteSizeSpan(size.roundToInt()), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
      }
      start = end
    }
    label.text = styled
  }
  fun recycle() { visual.recycle(); amountViews.forEach { it.recycle() }; description.setOnClickListener(null); actions.removeAllViews() }
  fun dispose() { recycle(); visual.dispose(); amountViews.forEach { it.dispose() } }

  private inner class AmountView : LinearLayout(context) {
    private val leading = NativeListLeadingVisual(reactContext)
    private val labels = LinearLayout(context).apply { orientation = VERTICAL }
    private val primary = NativeListTextView(context)
    private val secondary = NativeListTextView(context)
    init {
      orientation = HORIZONTAL; gravity = Gravity.CENTER_VERTICAL
      labels.addView(primary); labels.addView(secondary)
      addView(leading); addView(labels, LayoutParams(0, -2, 1f))
    }
    fun bind(data: JSONObject, key: String, style: JSONObject, theme: JSONObject?, table: Boolean) {
      val visualData = data.optJSONObject("leading")
      leading.visibility = if (visualData == null) GONE else VISIBLE
      leading.layoutParams = LayoutParams(dp(20), dp(20)).apply { marginEnd = dp(6) }
      if (visualData != null) leading.bind(visualData, JSONObject().put("width", 20).put("height", 20), key, theme, false, sourceScale) else leading.recycle()
      val primaryStyle = JSONObject(style.optJSONObject("primaryAmount")?.toString() ?: "{}")
      if (!primaryStyle.has("alignment")) primaryStyle.put("alignment", if (table) "start" else "end")
      text(primary, data.optString("text"), primaryStyle, 16f, color(data.optString("tone"), theme), 24, "medium")
      applySegments(primary, data.optJSONArray("textSegments"), primaryStyle.optDouble("fontSize", 16.0))
      val secondaryStyle = JSONObject(style.optJSONObject("secondaryAmount")?.toString() ?: "{}")
      if (!secondaryStyle.has("alignment")) secondaryStyle.put("alignment", if (table) "start" else "end")
      text(secondary, data.optString("secondaryText"), secondaryStyle, 14f, color("secondary", theme), 20)
      applySegments(secondary, data.optJSONArray("secondaryTextSegments"), secondaryStyle.optDouble("fontSize", 14.0))
    }
    fun recycle() { leading.recycle() }
    fun dispose() { leading.dispose() }
  }
}
